### PR TITLE
Make clang report invalid target versions.

### PR DESCRIPTION
**Pull request by ZijunZhaoCCK**
Clang always silently ignores garbage target versions and this makes debug harder. So clang will report when target versions are invalid.

### CODE REVIEW

File: llvm/lib/TargetParser/Triple.cpp
Comment (enh-google) [inline llvm/lib/TargetParser/Triple.cpp:1205]: maybe better to just go with `"version " << Name << " is invalid"`, rather than try to give examples, since they're so different between Android and iOS/macOS?

File: llvm/lib/TargetParser/Triple.cpp
Comment (pirama-arumuga-nainar) [inline llvm/lib/TargetParser/Triple.cpp:1204]: Also include the triple in the error message: `version _ in triple _ is invalid".

**Comment by pirama-arumuga-nainar** [comment]
@llvm/clang-vendors since this change can affect downstream components.

Can we do this change in two steps?
1. Since we want this error for Android, let's make a PR that restricts the error only to Android.
2. Make a follow-up PR to report error for all environments.  We can revert/iterate on this PR until all downstream errors are handled.

**Comment by rnk** [comment]
This change requires tests

**Comment by tru** [comment]
This needs to be cleaned up before it can be merged. There needs to be only one commit in this PR without the merge commits from main.

**Comment by tru** [comment]
It also needs tests.

**Comment by zmodem** [comment]
Thanks for looking at this. Silently ignoring bad inputs is usually not a good idea.

However, it would seem better to emit a proper diagnostic from the driver rather than calling exit in getOSDefines(). That way the regular diagnostic mechanisms can be used to decide what to do with it (treat as error, ignore, etc.).

**Comment by MaskRay** [comment]
In addition, calling `exit` from library code is probably not good. The error can be surfaced by a very top-level component (e.g. driver as @zmodem mentioned).

File: clang/lib/Driver/ToolChain.cpp
Comment (ZijunZhaoCCK) [inline clang/lib/Driver/ToolChain.cpp:692]: test in `clang/unittests/Driver/ToolChainTest.cpp` needed

File: clang/unittests/Driver/DriverTest.cpp
Comment (ZijunZhaoCCK) [inline clang/unittests/Driver/DriverTest.cpp:9]: Todo for 2024:
1. add more test cases
2. remove unused header files
3. depend on further code reviews

File: clang/lib/Driver/Driver.cpp
Comment (pirama-arumuga-nainar) [inline clang/lib/Driver/Driver.cpp:1438]: Check for `Triple.getEnvironmentVersion().empty()` instead of `.getMajor() == 0`.  In non-Android contexts, a major version of `0` may be valid.

**Review by pirama-arumuga-nainar — APPROVED** [review]
Please also wait for @MaskRay's LGTM before merging.

File: clang/include/clang/Basic/DiagnosticDriverKinds.td
Comment (MaskRay) [inline clang/include/clang/Basic/DiagnosticDriverKinds.td:791]: No capitalization and drop the full stop to be consistent with other messages/convention.

File: clang/include/clang/Basic/DiagnosticDriverKinds.td
Comment (MaskRay) [inline clang/include/clang/Basic/DiagnosticDriverKinds.td:790]: You can name it `err_triple_version_invalid` in case other OSes want to do the same checking.

**Review by MaskRay — APPROVED** [review]

File: clang/include/clang/Basic/DiagnosticDriverKinds.td
Comment (jrtc27) [reply clang/include/clang/Basic/DiagnosticDriverKinds.td:791]: * Elsewhere we say "target triple"
* Repeating the version seems a bit ugly?

Comment (MaskRay) [reply clang/include/clang/Basic/DiagnosticDriverKinds.td:791]: Perhaps `version '%0' in target triple '%1' is invalid`

**Pull request merged**
